airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultGetSpecWorker.java

```diff
@@ -59,39 +55,37 @@ public ConnectorJobOutput run(final JobGetSpecConfig config, final Path jobRoot)
     ApmTraceUtils.addTagsToTrace(Map.of(JOB_ROOT_KEY, jobRoot, DOCKER_IMAGE_KEY, config.getDockerImage()));
     try {
       process = integrationLauncher.spec(jobRoot);
+      final int exitCode = process.exitValue();
+      final String exitCodeMessage = String.format("Spec job subprocess finished with exit code %s", exitCode);
+      LOGGER.debug(exitCodeMessage);
 
+      final ConnectorJobOutput jobOutput = new ConnectorJobOutput().withOutputType(OutputType.SPEC);
       LineGobbler.gobble(process.getErrorStream(), LOGGER::error);
 
-      final Map<Type, List<AirbyteMessage>> messagesByType;
-      try (final InputStream stdout = process.getInputStream()) {
-        messagesByType = streamFactory.create(IOs.newBufferedReader(stdout))
-            .collect(Collectors.groupingBy(AirbyteMessage::getType));
-
-        // todo (cgardens) - let's pre-fetch the images outside of the worker so we don't need account for
-        // this.
-        // retrieving spec should generally be instantaneous, but since docker images might not be pulled
-        // it could take a while longer depending on internet conditions as well.
-        WorkerUtils.gentleClose(process, 30, TimeUnit.MINUTES);
-      }
+      final Map<Type, List<AirbyteMessage>> messagesByType = WorkerUtils.getMessagesByType(process, streamFactory, 30);
 
       final Optional<ConnectorSpecification> spec = messagesByType
           .getOrDefault(Type.SPEC, new ArrayList<>()).stream()
           .map(AirbyteMessage::getSpec)
           .findFirst();
 
-      final int exitCode = process.exitValue();
-      if (exitCode == 0) {
-        if (spec.isEmpty()) {
-          throw new WorkerException("integration failed to output a spec struct.");
-        }
+      final Optional<FailureReason> failureReason = WorkerUtils.getJobFailureReasonFromMessages(OutputType.SPEC, messagesByType);
+      failureReason.ifPresent(jobOutput::setFailureReason);
 
-        return new ConnectorJobOutput().withOutputType(OutputType.SPEC).withSpec(spec.get());
+      if (spec.isPresent()) {
+        jobOutput.setSpec(spec.get());
       } else {
-        return WorkerUtils.getJobFailureOutputOrThrow(
-            OutputType.SPEC,
-            messagesByType,
-            String.format("Spec job subprocess finished with exit code %s", exitCode));
+        if (failureReason.isEmpty()) {
+          final String stderr = WorkerUtils.getStdErrFromErrorStream(process.getErrorStream());
+          if (stderr.isEmpty()) {
+            throw new WorkerException("Integration failed to output a spec struct and did not output a failure reason.");
+          } else {
+            throw new WorkerException("Integration failed to output a spec struct and did not output a failure reason:\n" + stderr);
+          }
+        }
```

Comment:
Nit: here you indicate that there wasn't a failure reason for both cases, which seems correct here - but in the other two jobs, it's left out for both, or only included where there is no stderr output. 

Also wondering if it's worth combining these back into a function like before (but not returning) - something like `throwOnEmptyFailureReason("Integration failed to output a spec struct and did not output a failure reason.")`, since they all use the same code except for the included error message

Reply:
I combined this logic in a utility function ✅ .

> here you indicate that there wasn't a failure reason for both cases, which seems correct here - but in the other two jobs, it's left out for both, or only included where there is no stderr output.

I'm not sure I understand this comment. I believe the logic is similar across the three classes because I implemented the utility function.

Reply:
I figured you mean the error message :D